src/components/views/messages/MImageBody.js

```diff
@@ -105,24 +105,46 @@ module.exports = React.createClass({
         return MatrixClientPeg.get().mxcUrlToHttp(content.url, 800, 600);
     },
 
-    render: function() {
-        var TintableSvg = sdk.getComponent("elements.TintableSvg");
+    componentDidMount: function() {
+        this.dispatcherRef = dis.register(this.onAction);
+        this.fixupHeight();
+    },
+
+    componentWillUnmount: function() {
+        dis.unregister(this.dispatcherRef);
+    },
+
+    onAction: function(payload) {
+        if (payload.action === "timeline_resize") {
+            this.fixupHeight();
+        }
+    },
+
+    fixupHeight: function() {
         var content = this.props.mxEvent.getContent();
-        var cli = MatrixClientPeg.get();
 
         var thumbHeight = null;
-        if (content.info) thumbHeight = this.thumbHeight(content.info.w, content.info.h, 800, 600);
+        var timelineWidth = this._body.offsetWidth;
+        var maxHeight = 600; // let images take up as much width as they can so long as the height doesn't exceed 600px.
+        // the alternative here would be 600*timelineWidth/800; to scale them down to fit inside a 4:3 bounding box
+
+        //console.log("trying to fit image into timelineWidth of " + this._body.offsetWidth + " or " + this._body.clientWidth);
+        if (content.info) thumbHeight = this.thumbHeight(content.info.w, content.info.h, timelineWidth, maxHeight);
+        this._image.style.height = thumbHeight + "px";
+    },
 
-        var imgStyle = {};
-        if (thumbHeight) imgStyle['maxHeight'] = thumbHeight;
+    render: function() {
+        var TintableSvg = sdk.getComponent("elements.TintableSvg");
+        var content = this.props.mxEvent.getContent();
+        var cli = MatrixClientPeg.get();
 
         var thumbUrl = this._getThumbUrl();
         if (thumbUrl) {
             return (
-                <span className="mx_MImageBody">
+                <span className="mx_MImageBody" ref={(c) => this._body = c}>
                     <a href={cli.mxcUrlToHttp(content.url)} onClick={ this.onClick }>
-                        <img className="mx_MImageBody_thumbnail" src={thumbUrl}
-                            alt={content.body} style={imgStyle}
```

Comment:
we don't know what the right height is at that point because the DOM 
hasn't laid out enough to know what the available width is (and thus 
what the correct height is). i think?

On 24/03/2016 14:05, Richard van der Hoff wrote:
>
> In src/components/views/messages/MImageBody.js 
> <https://github.com/matrix-org/matrix-react-sdk/pull/248#discussion_r57320589>:
>
> >                      <a href={cli.mxcUrlToHttp(content.url)} onClick={ this.onClick }>
> > -                        <img className="mx_MImageBody_thumbnail" src={thumbUrl}
> > -                            alt={content.body} style={imgStyle}
>
> do we not want to initialise this with the right height, rather than 
> wait for the DidMount?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/matrix-org/matrix-react-sdk/pull/248/files/8b67a1059d1a40b3286fed17ebdbb3e8d327de94#r57320589>
>